Added draft blog post for buildpacks rebase (#283)
cncf/tag-env-sustainability: website/content/blog/2023-11-reduce-reuse-rebase-buildpacks.md

---
title: "Reduce, Reuse, Rebase: Sustainable Containers with Buildpacks"

> description: Buildpacks are the greenest choice for container builds

Reviewer: I would like to remove claims like "the greenest choice."  Such assertions should not be done without measurements, and there may be other "greener" container builds.

Author: I suggested some different language in a comment below

Author: Fixed in e0aea31

slug: 2023-11-reduce-reuse-rebase-buildpacks
weight: 1
date:   2023-11-17 00:00:00 +0000
author: Joe Kutner
---

> Container builds can be very wasteful. Every operating system update, new dependency version, and toolchain upgrade results in copious amounts of energy used to build and rebuild our container images; often unnecessarily. It can be costly at scale, which is why Cloud Native Buildpacks were designed to perform builds as efficiently as possible.

Reviewer 2: Thoughts
* Maybe add one sentence at the start to get the reader into the mindset. (ofc we expect that the reader knows docker etc.)
* `Container builds can be very wasteful.` does not actually say much. The next sentence says more.
* `which is why Cloud Native Buildpacks were designed to perform builds as efficiently as possible.` this is quite long and does not really say anything. Every tool wants so be performant. What comes in the next paragraphs rebasing etc. are interesting because these are the concepts that differ.. 
* To me, a history perspective is missing here. Some details into docker are described a bit later, but I sense that there is something missing here. (feel free to disagree, ofc.)
* Perhaps you want to merge both paragraphs. I am not sure why to break at this point.

Author: How about this?

>  which is why Cloud Native Buildpacks were designed to perform full builds only when a rebuild is actually required.

Re: history perspective, I worry that might clutter the intro. But maybe it is necessary to set context. Are you thinking I should add some of the history of buildpacks (ex. they pre-date `Dockerfile`). Also, it may be worth clarifying that the choice is not between Docker and Buildpacks, but rather `Dockerfile` and Buildpacks (buildpacks image work with docker and otehr container runtimes).

For the intro sentence, I'm not sure what you mean by "get the reader into the mindset". Can you elaborate?

Reviewer 2: > How about this?
> [...]

👍 

> I worry that might clutter the intro. But maybe it is necessary to set context

Yeah, the history part should come later, but for the flow and to get into the context it would consider adding / teasing it in the intro. 

> Also, it may be worth clarifying that the choice is not between Docker and Buildpacks, but rather Dockerfile and Buildpack

Yes! I would take this in. You could think about using `Containerfile` instead of `Dockerfile`. 

> For the intro sentence, I'm not sure what you mean by "get the reader into the mindset"

Essentially, my comment is about setting the context. The blogs we have are very different in terms of content. Just to make it easier to change context. Buildpacks is used for XYZ.

> Buildpacks are the greenest choice for container builds because the rebaseable layers and advanced caching mechanisms they provide can prevent many images from being rebuilt at all. That's a big shift from many cloud native technologies that assume unlimited resources are available in the cloud.

Reviewer: Again, please remove such claims as "the greenest choice."

Author: Certainly. What about this?

> Buildpacks are an environmentally friendly choice for container builds ...

<p class="mt-5 mb-5"><img src="/images/blogs/2023-11-reduce-reuse-rebase-buildpacks/cnb-green-logo.png" alt="CNB logo with recycling"></p>

## On the Environmental Impact of Cloud Native

Prior to the emergence of the cloud native ecosystem and widespread use of container images, we deployed our applications in servers built from machine images that were updated infrequently and on a different cadence from the application itself.

Today, many applications are coupled to the operating system and its packages because they use a Dockerfile to define their container images. As a result, those images frequently need rebuilds to apply patches to OS-level components, or simply to update tools that aren't even used by the application. Even worse, the layer caching mechanism imposed by Dockerfile forces us to frequently rebuild layers that don't even need to be rebuilt.

> The cloud-native ecosystem has brought great productivity and operational improvements to software development. But we've lost sight of how wasteful some of those technologies can be.

Reviewer 3: > But we've lost sight of how wasteful some of those technologies can be.
@jkutner Can you elaborate what do you mean by wasteful here?

Author: What do you think about something like this?

> The cloud-native ecosystem has brought great productivity and operational improvements to software development. But we've lost sight of how wasteful some of those technologies can be. The relative ease and low cost of cloud infrastructure encourages us to solve problems by spinning up new resources instead of optimizing for the efficient use of our resources. We see this with container builds using Dockerfile, but also other parts of CI/CD, where it's sometimes easier to re-run expensive processes.

> Buildpacks, on the other hand, have been designed to work at a scale (i.e. tens millions of images) where being wasteful has real costs. That's why the Buildpacks rebase mechanism requires almost no resources.

Reviewer: ```suggestion
Buildpacks, on the other hand, have been designed to work at a scale (i.e. tens millions of images) where being wasteful has real costs. That's why the Buildpacks rebase mechanism requires minimal resources.
```

## Reduce, Reuse, Rebase

> Container images built from a Dockerfile require a full build when a new operating system update is available, even if your app doesn't required a re-compile or re-install to work with the update (i.e. the update is [ABI compatible](https://en.wikipedia.org/wiki/Application_binary_interface)). 

Reviewer 3: nit: required -> require

<p class="mt-5 mb-5"><img src="/images/blogs/2023-11-reduce-reuse-rebase-buildpacks/cnb-new-base.png" alt="Updating the operating system of a container image requires a rebuild if you're using Dockerfile"></p>

> Updating the operating system of a container image requires a rebuild if you're using Dockerfile.But this isn't the case with Buildpacks.

Reviewer: ```suggestion
Updating the operating system of a container image requires a rebuild if you're using Dockerfile.  This in no longer the case with Buildpacks.
```

Author: How about something like "This is not the case when using Buildpacks"?

I feel like "no longer" implies that Buildpacks are a new thing, but that are actually older than Docker itself.

Author: Fixed in c53ffe9

<p class="mt-5 mb-5"><img src="/images/blogs/2023-11-reduce-reuse-rebase-buildpacks/cnb-rebase.png" alt="Buildpacks rebase constructs a new container image using existing layers, without the need for build"></p>

Rebase allows app developers or operators to rapidly update an app image when its run image has changed. By using image layer rebasing, this command avoids the need to fully rebuild the app.

> Buildpacks rebase constructs a new container image using existing layers, without the need for build.At its core, image rebasing is a simple process. By inspecting an app image, rebase can determine whether or not a newer version of the app's base image exists (either locally or in a registry). If so, rebase updates the app image's layer metadata to reference the newer base image version. This is essentially a an operation that edits a JSON file. It takes milliseconds and uses very little compute resources. 

Reviewer: ```suggestion
Buildpacks rebase constructs a new container image using existing layers, without the need for build. At its core, image rebasing is a simple process. By inspecting an app image, rebase can determine whether or not a newer version of the app's base image exists (either locally or in a registry). If a newer version exists, rebase updates the app image's layer metadata to reference the newer base image version. This is essentially a an operation that edits a JSON file. It takes milliseconds and uses very little compute resources. 
```

You can [learn more about rebase in the Buildpacks documentation](https://buildpacks.io/docs/concepts/operations/rebase/). But rebase isn't the only Buildpacks mechanism that's more sustainable that Dockerfile builds. Buildpacks can cache build artifacts to enable incremental compiliation, and other resource saving techniques. These cache layers won't always be discarded when you do require a re-build, like they would with Dockerfile builds.

## Be As Green As Your Unit Tests

> Container builds are in no way the biggest offender when it comes to the environmental impact of software. Electricity needed to mine bitcoin is [more than used by entire countries](https://www.theguardian.com/technology/2021/feb/27/bitcoin-mining-electricity-use-environmental-impact). But the growth of software that uses cryptographic techniques has brought new awareness to how our code affects the world around us. That's a good thing.

Reviewer: ```suggestion
Container builds are not the biggest offender when it comes to the environmental impact of software. Electricity needed to mine bitcoin is [more than used by entire countries](https://www.theguardian.com/technology/2021/feb/27/bitcoin-mining-electricity-use-environmental-impact), but the growth of software that uses cryptographic techniques has brought new awareness to how our code affects the world around us. That's a good thing.
```

> We have a responsibility to think about the waste and sustainability of the software we produce. The code we write has an impact on the world, and our choices matter.

Reviewer: ```suggestion
We have a responsibility to think about minimizing the required resources of the software we produce. The code we write has an impact on the world, and our choices matter.
```

> To learn more about the relationship between open source software development any the environment, visit the [Environmental Sustainability Technical Advisory Group (TAG)](https://tag-env-sustainability.cncf.io/)

Reviewer 3: nit: any -> and

Author: Fixed in 4dc17b6
